### ISSUE TITLE
Features' view observers should be scoped

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-engine#1171]
Right now if we have for instance two features – lists and tables, they both may listen to the <kbd>Enter</kbd> key. Lists will handle it by creating a new item, while tables by moving the caret to the next cell.

However, what if a list is created inside a table? Then, conflict starts. Depending on the order in which these listeners were added, only the first one will handle the event, so it may happen that pressing <kbd>Enter</kbd> inside a list will move the caret to the next cell.

We deal with that in two ways (depending on the situation):

* By using listener priorities.
* By implementing the table's listener in such a way that it checks whether the selection is really inside a table cell and not somewhere deeper.

Both solutions have flaws and lead to complicated code.

A better solution would be if lists could listen to the `keydown` event on every `<li>` element. Like an early return in a function – catching the event early would prevent it from bubbling up to the table (which would handle the event otherwise).

However, this won't work natively – the `keydown` event will not be fired on a list, nor on a table – it will be fired on the element which has `contentEditable=true` attribute (on the editing host). And then it will bubble up to the `<body>`.

You can check it here: https://jsfiddle.net/nhtLnuoh/1/.

Since native bubbling doesn't work here, what would? Perhaps we're able to implement a system where you define a listener's context (like the `selector` in [jQ's `delegate()`](http://api.jquery.com/delegate/)). Then, the listeners will be filtered by whether the context matches and executed in an order from the longest path to shortest.

E.g. if you defined two listeners:

* `'keydown'`, ctx: `'table'`
* `'keydown'`, ctx: `'li'`
* `'keydown'`, ctx: `'a'`

Then if an event was fired in `table > tr > td > li > {foo}`:

1. first, the listener for `'li'` would be called
2. then, for `'table'`
3. and the one for `'a'` would not be called at all

Simple event delegation mechanism. It will make listeners more contextual and hence code better organised and safer (no more table's listener kicking in before list's or priorities hell).

**Comment by jodator** [comment]
I like this idea. Would make code in other parts shorter since I guess that such check in `enter` key lister in list:
https://github.com/ckeditor/ckeditor5-list/blob/b1d47eec86f7eed37e6ba820bfb18ded9855c088/src/list.js#L52-L56
would be at least shorter by checking parent.
This has also a potential to reduce number of listeners called on each 'enter' to a number of those interested.

**Comment by pjasiun** [comment]
The concept is interesting, but first, we should consider how many cases we have like this. The collision between lists and tables can be solved as you mentioned, and if it's the only case we have in mind it might be simply not worth to bring another abstraction. The current need for the refactoring proves us that it's better to introduce a new mechanism after gathering a few real cases. But again: the concept sounds interesting. :)

**Comment by oleq** [comment]
> The concept is interesting, but first, we should consider how many cases we have like this.

Another example is a link in a list. ATM we have a high priority <kbd>Tab</kbd> listener in Link which moves the focus to the link editing balloon instead of indenting the list because accessibility is the priority at this point. 

A scoped listener would do the trick here because the link is deeper in the tree. I'm not sure whether this should be handled on the engine's view level or in `editor.keystrokes`, though.

**Comment by pjasiun** [comment]
What was not clear is if this scoping should be done on the view or on the model. Observables know nothing about the model. If it supposes to be part of them it should not touch model. Also, @oleq case needs to work on the view.

On the other hand, tables vs list case should use the model, if it supposes to be done right. If we will assume that the fact that there is `li` or `td` in the view, means that it will be mapped to the `listItem` or `tableCell` for sure and we do not check it later, it will not be safe. For such problem, much better would be some util to fire events on the model element based on the model selection.

This is why I believe we need to gather some cases. We will see the common part and learn what we really need.

**Comment by Reinmar** [comment]
Just to clarify this – this was brought to us by CKEditor 4's users. The actual cases that we discussed were related to a really complex system with a rich structure of elements. Far more than what we have now for CKEditor 5 and even far more than what CKEditor 4 offers. So, I can't give you exact cases but I can imagine that some systems create those.

**Comment by Reinmar** [comment]
> What was not clear is if this scoping should be done on the view or on the model. Observables know nothing about the model. If it supposes to be part of them it should not touch model. Also, @oleq case needs to work on the view.

You're right. Most of the algorithms (including @oleq's one – cause why not?) should depend on the model... So the example I gave in my initial post is wrong – it should use model element names.

**However, this makes that delegation harder because it leaks from the view layer (where the event is being handled) to the model layer (which structure's matter).**
